### PR TITLE
[FABG-948] Add support for extra hosts at enrollment

### PR DESCRIPTION
**Pull request by vieirin**
Provide WithHosts option and add hosts to CSR info at enrollment request
This allows us to pass extra hosts for certificate enrollment, so it can be valid to
those hosts as well for its CN

Signed-off-by: Manoel <vieiranetoc@gmail.com>

### CODE REVIEW

**Comment by vieiramanoel** [comment]
Pls run codecov again, test added for WithHosts

**Review by alikic — CHANGES_REQUESTED** [review]
EnrollmentRequest should follow https://github.com/hyperledger/fabric-sdk-go/blob/b5424d1fa32172cca169bf9b8d549f956144ff5a/internal/github.com/hyperledger/fabric-ca/api/client.go#L54. Adding just host is not a good idea as it is conceivable one might want to add more from CSRInfo in the future. So the whole CSRInfo should be added to EnrollmentRequest instead of just host.

**Comment by vieiramanoel** [comment]
Integration tests are breaking, but I don't think it's a problem with the code itself, updating branch to see if it passes

**Comment by vieiramanoel** [comment]
@alikic any other changes?

File: pkg/client/msp/ca.go [hunk 1]
```diff
@@ -6,6 +6,8 @@ SPDX-License-Identifier: Apache-2.0
 
 package msp
 
+import "github.com/cloudflare/cfssl/csr"
```
Comment (alikic) [inline pkg/client/msp/ca.go]: We don't include third party structures in our high level API. This import shouldn't be here.

File: pkg/client/msp/ca.go [hunk 2]
```diff
@@ -195,3 +197,19 @@ type GetCAInfoResponse struct {
...
 	Version string
 }
+
+// CSRInfo is Certificate Signing Request (CSR) Information
+type CSRInfo struct {
+	CN           string
+	Names        []csr.Name
+	Hosts        []string
+	KeyRequest   *BasicKeyRequest
+	CA           *csr.CAConfig
```
Comment (alikic) [inline pkg/client/msp/ca.go]: This API shouldn't have dependencies on "github.com/cloudflare/cfssl/csr". CAConfig should be created here, and any mapping between this API and third party should be hidden behind the API. While we have a freedom to design our API structures independently from the third party, it is useful to keep them close, as we keep extending support for Fabric CA feature

**Review by alikic — CHANGES_REQUESTED** [review]

File: pkg/client/msp/ca.go (hunk 2 above)
Comment (vieiramanoel) [reply pkg/client/msp/ca.go]: I strongly disagree, to create a new struct and re-map it as done in CSR struct seems a bad thing to me, CA depends on cloudflare structure, this dependency is already packaged with sdk code, btw. 
I don't see how creating same structs and "convert this type to that type" functions, where those types are the same
what do you think

Comment (alikic) [reply pkg/client/msp/ca.go]: Neither fabric-ca nor cloudflare/cfssl are used directly in the exposed API, all uses are internal. I see that we do use cloudflare/cfssl in a couple of tests, perhaps we shouldn't, we can easily change that. Usage of cloudflare/cfssl internally is an implementation detail, shouldn't be in public API.

You wanted to add only hosts to CSR info, and my thinking was, if we keep adding the missing pieces of Fabric CA support, we will likely end up with structures similar to what the CA is using. This is why I thought we should introduce CSRInfo, to group hosts and future (currently missing) elements together in a way it makes sense. CSRInfo could initially have only hosts, and whatever you currently need. Ideally, all additions should have proper integration tests.

**Comment by chanioxaris** [comment]
Any updates on this issue?

**Comment by alikic** [comment]
Here is where we stand - the feature is useful, but the implementation has two issues. It introduces a third party dependency in the public API, and it is missing tests. It makes sense to group all parameters relevant to the creation of a CSR into a structure (perhaps initially having only hosts in the structure), which could be called CSRInfo, but no third party structures are allowed in the public API.

**Comment by vieiramanoel** [comment]
I'm going back to this issue, @alikic this test should be written in integration tests too, right? 
On unit tests return is always empty for hosts entry on returned certificate.

**Comment by vieiramanoel** [comment]
Integration tests are broken again
updating with last master hope it fix

**Comment by vieiramanoel** [comment]
Done :)

File: test/integration/pkg/client/msp/enrollment_test.go [hunk 3]
```diff
@@ -304,8 +305,124 @@ func TestEnrollWithProfile(t *testing.T) {
...
+		Type: IdentityTypeUser,
+		// Affiliation is mandatory. "org1" and "org2" are hardcoded as CA defaults
+		// See https://github.com/hyperledger/fabric-ca/blob/release/cmd/fabric-ca-server/config.go
+		Affiliation: "org2",
+	})
+	if err != nil {
+		t.Fatalf("Registration failed: %s", err)
+	}
+
+	extraHosts := []string{"localhost"}
```
Comment (alikic) [inline test/integration/pkg/client/msp/enrollment_test.go]: Need to test with more values in extraHosts, it should add all of them to the cert....

File: test/integration/pkg/client/msp/enrollment_test.go [hunk 4]
```diff
@@ -304,8 +305,124 @@ func TestEnrollWithProfile(t *testing.T) {
...
+		t.Fatalf("Enroll failed: %s", err)
+	}
+
+	// Get the new user's signing identity
+	si, err := mspClient.GetSigningIdentity(username)
+	if err != nil {
+		t.Fatalf("GetSigningIdentity failed: %s", err)
+	}
+
+	has, err := hasHost(si, extraHosts[0])
```
Comment (alikic) [inline test/integration/pkg/client/msp/enrollment_test.go]: .... and then verify that all the requested hosts are in the cert.

File: test/integration/pkg/client/msp/enrollment_test.go [hunk 5]
```diff
@@ -304,8 +305,124 @@ func TestEnrollWithProfile(t *testing.T) {
...
+		t.Fatalf("Reenroll failed: %s", err)
+	}
+
+	// Get the new user's signing identity
+	si, err = mspClient.GetSigningIdentity(username)
+	if err != nil {
+		t.Fatalf("GetSigningIdentity failed: %s", err)
+	}
+
+	has, err = hasHost(si, extraHosts[0])
```
Comment (alikic) [inline test/integration/pkg/client/msp/enrollment_test.go]: .... and then verify that all the requested hosts are in the cert.

**Review by alikic — CHANGES_REQUESTED** [review]

File: test/integration/pkg/client/msp/enrollment_test.go (hunk 5 above)
Comment (alikic) [reply test/integration/pkg/client/msp/enrollment_test.go]: otherwise looks good

File: test/integration/pkg/client/msp/enrollment_test.go (hunk 3 above)
Comment (vieiramanoel) [reply test/integration/pkg/client/msp/enrollment_test.go]: Can this comparison be O(n²) in test or should it be optimized with a map-based search? 
I mean can I compare all requested hosts with all received hosts or convert requested hosts to a map first to compare? Does it makes difference? Just to write it right from start haha

Comment (alikic) [reply test/integration/pkg/client/msp/enrollment_test.go]: Haha, you can implement comparison any way you want, I am just looking for having at least two values there, just to have a little more confidence that the mapping works correctly through the layers all the way to the CA itself.

Comment (vieiramanoel) [reply test/integration/pkg/client/msp/enrollment_test.go]: Ok. Understood.

**Review by alikic — APPROVED** [review]

**Pull request merged**